server/src-lib/Hasura/Server/Auth/JWT.hs

```diff
@@ -400,13 +400,14 @@ instance J.FromJSON JWTConfig where
       (Nothing, Nothing) -> fail "key and jwk_url both cannot be empty"
       (Just _, Just _)   -> fail "key, jwk_url both cannot be present"
       (Just rawKey, Nothing) -> do
-        key <- parseKey keyType rawKey
+        keyType' <- onNothing keyType $ fail "type of key not provided"
+        key <- parseKey rawKey keyType'
```

Comment:
In the PR description, you write:

> The `type` key is only used to parse the type of `HS` algorithms so that we can throw a short key error. Nothing else. So it might be worth removing it alltogether.

So it sounds like we only do anything with it right here. In that case, it sounds like we might as well drop it from the `JWTConfig` type, and we can move the `keyType <- o J..: "type"` statement directly into this nested `do` block:

```haskell
(Just rawKey, Nothing) -> do
  keyType <- o J..: "type"
  key <- parseKey rawKey keyType
```

Then we don’t have to bother with `J.:?` only to assert its presence later in this branch.

Reply:
That does sound like it reduces some code but it also reduces visibility of the what the config can have at the top level. I personally feel it's more readable if everything is at one place.

So, we can remove the `type` field from JWTConfig but let's parse it at the top?

Reply:
I disagree. The purpose of aeson’s monadic `Parser` abstraction is to give you the ability to use control flow to make decisions about parsing. Taking advantage of that is the idiomatic thing to do here; parsing something outside of a branch that’s only used inside the branch would be more confusing to me than parsing it locally!